### PR TITLE
Doc tokenizer

### PR DESCRIPTION
**Pull request by sgugger**
Improve the documentation of tokenizers, following what was done for the models last week, mainly:
- make sure all docstrings of public functions are properly formatted for sphinx
- make sure all args are properly documented
- add or fix type hints wherever necessary

The methods/classes that are not in the main `__init__` are all in the page `internal/tokenization_utils.html`. I had added `SpecialTokensMixin` in the __init__ of transformers a while ago to easily document it, but it can be removed from there now if we want.

I rewrote a few dosctrings here and there so pinging @n1t0 and @mfuntowicz to make sure I didn't write anything bad.

[Preview](https://65719-155220641-gh.circle-artifacts.com/0/docs/_build/html/main_classes/tokenizer.html) of the tokenization page.
[Preview](https://65719-155220641-gh.circle-artifacts.com/0/docs/_build/html/internal/tokenization_utils.html) of the tokenization utils page.

### CODE REVIEW

File: src/transformers/tokenization_utils.py [hunk 1]
```diff
@@ -675,16 +692,40 @@ def convert_ids_to_tokens(
     def _convert_id_to_token(self, index: int) -> str:
         raise NotImplementedError
 
-    def convert_tokens_to_string(self, tokens: List[str]) -> str:
-        """ Converts a sequence of tokens (string) in a single string.
-            The most simple way to do it is ' '.join(self.convert_ids_to_tokens(token_ids))
-            but we often want to remove sub-word tokenization artifacts at the same time.
+    def convert_tokens_to_string(self, tokens: List[int]) -> str:
```
Comment (sgugger) [inline src/transformers/tokenization_utils.py:695]: This method add a typehint and a docstring that is not compatible with the way the result is computed (`convert_ids_to_tokens` takes token ids, not tokens). Should the argument `tokens` be renamed `token_ids` or `ids`?

Comment (LysandreJik) [reply src/transformers/tokenization_utils.py:695]: I'd go with `token_ids`

File: docs/source/main_classes/tokenizer.rst [hunk 2]
```diff
@@ -1,17 +1,40 @@
...
+
+- Tokenizing (splitting strings in sub-word token strings), converting tokens strings to ids and back, and
+  encoding/decoding (i.e., tokenizing and converting to integers).
+- Adding new tokens to the vocabulary in a way that is independent of the underlying structure (BPE, SentencePiece...).
+- Managing special tokens (like mask, beginning-of-sentence, etc.): adding them, assigning them to attributes in the
+  tokenizer for easy access and making sure they are not split during tokenization.
+
+:class:`~transformers.BatchEncoding` holds the output of the tokenizer's encoding methods (``__call__``,
+``encode_plus`` and ``batch_encode_plus``) and is derived from a Python dictionary. When the tokenizer is a pure python
+tokenizer, this class behave just like a standard python dictionary and hold the various model inputs computed by these
```
Comment (LysandreJik) [inline docs/source/main_classes/tokenizer.rst:32]:
```suggestion
tokenizer, this class behaves just like a standard python dictionary and holds the various model inputs computed by these
```

**Review by LysandreJik — APPROVED** [review]
Very nice; good work!

File: docs/source/main_classes/tokenizer.rst [hunk 3]
```diff
@@ -1,17 +1,40 @@
 Tokenizer
 ----------------------------------------------------
 
-A tokenizer is in charge of preparing the inputs for a model. The library comprise tokenizers for all the models. Most of the tokenizers are available in two flavors: a full python implementation and a "Fast" implementation based on the Rust library `tokenizers`. The "Fast" implementations allows (1) a significant speed-up in particular when doing batched tokenization and (2) additional methods to map between the original string (character and words) and the token space (e.g. getting the index of the token comprising a given character or the span of characters corresponding to a given token). Currently no "Fast" implementation is available for the SentencePiece-based tokenizers (for T5, ALBERT, CamemBERT, XLMRoBERTa and XLNet models).
+A tokenizer is in charge of preparing the inputs for a model. The library comprise tokenizers for all the models. Most
```
[the remark this answers is not in the document]
Comment (sgugger) [reply docs/source/main_classes/tokenizer.rst:4]: Yeah contains is probably better-suited.

File: src/transformers/tokenization_utils.py (hunk 1 above)
Comment (sgugger) [reply src/transformers/tokenization_utils.py:695]: Actually looking at all the subclasses, it's this implementation that is wrong: all the subclasses get tokens and return strings, so the base class should just join on space and not call `convert_ids_to_tokens`.
@thomwolf or @n1t0, could you confirm?

File: src/transformers/tokenization_utils_base.py [hunk 4]
```diff
@@ -212,20 +241,39 @@ def items(self):
...
-
-        Returns: List[EncodingFast] or None if input was tokenized through Python (i.e. not fast) tokenizer
+        :obj:`Optional[List[tokenizers.Encoding]]`: The list all encodings from the tokenization process.
+        Returns :obj:`None` if the input was tokenized through Python (i.e., not a fast) tokenizer.
         """
         return self._encodings
 
     def tokens(self, batch_index: int = 0) -> List[str]:
+        """
+        Return the list of tokens at a given index for a fast tokenizer.
```
[the remark this answers is not in the document]
Comment (thomwolf) [reply src/transformers/tokenization_utils_base.py:251]: `index` => ` batch index`

[the remark this answers is not in the document]
Comment (thomwolf) [reply src/transformers/tokenization_utils_base.py:251]: `for a fast tokenizer` => `(only works for the output of fast tokenizers)`

File: src/transformers/tokenization_utils_base.py [hunk 5]
```diff
@@ -682,24 +770,27 @@ def add_special_tokens(self, special_tokens_dict: Dict[str, Union[str, AddedToke
...
-            new_tokens: string or list of string or :class:`~transformers.AddedToken`. Each string is a token to add.
-                Tokens are only added if they are not already in the vocabulary. AddedToken wrap a string token to
-                let you personnalize it's behavior (Whether this token should only match against single word, whether
-                this token should strip all potential whitespaces on the left side, Whether this token should strip
-                all potential whitespaces on the right side...).
-            special_token: can be used to specify if the token is a special token. This mostly change the normalization
-                behavior (special tokens like CLS or [MASK] are usually not lower-cased for instance)
+            new_tokens (:obj:`str`, :obj:`tokenizers.AddedToken` or a list of `str` or :obj:`tokenizers.AddedToken`):
+                Tokens are only added if they are not already in the vocabulary. :obj:`tokenizers.AddedToken` wraps a
+                string token to let you personnalize it's behavior: whether this token should only match against single
```
Comment (thomwolf) [inline src/transformers/tokenization_utils_base.py:783]:
```suggestion
                string token to let you personalize its behavior: whether this token should only match against a single
```

File: src/transformers/tokenization_utils_base.py [hunk 6]
```diff
@@ -1388,35 +1648,43 @@ def save_pretrained(self, save_directory) -> Tuple[str]:
...
-    ):
+    ) -> List[int]:
         """
-        Converts a string in a sequence of ids (integer), using the tokenizer and vocabulary.
+        Converts a string to a sequence of ids (integer), using the tokenizer and vocabulary.
 
         Same as doing ``self.convert_tokens_to_ids(self.tokenize(text))``.
 
+        .. warning::
+            This method is deprecated, ``__call__`` should be used instead.
```
Comment (thomwolf) [inline src/transformers/tokenization_utils_base.py:1677]: `encode` is still useful if you want a simple encoding of a string (without all the outputs) ie. `str -> List[int]`
I think we can remove the deprecation message.

**Review by thomwolf — APPROVED** [review]
Very nice!

**Pull request merged**